File: packages/shared/src/physics/body.ts
import { Vec2, vec2 } from './vector.js';

export interface Circle {
  position: Vec2;
  velocity: Vec2;
  radius: number;
  mass: number;
  invMass: number;
  restitution: number;
  friction: number;
  isStatic: boolean;
}

export interface Line {
  p1: Vec2;
  p2: Vec2;
  restitution: number;
}

export function createCircle(options: {
  position: Vec2;
  radius: number;
  mass?: number;
  restitution?: number;
  friction?: number;
  isStatic?: boolean;
  velocity?: Vec2;
}): Circle {
  const mass = options.mass ?? 1;
  const isStatic = options.isStatic ?? false;
  return {
    position: { ...options.position },
    velocity: options.velocity ? { ...options.velocity } : vec2(0, 0),
    radius: options.radius,
    mass: mass,
    invMass: isStatic ? 0 : 1 / mass,
    restitution: options.restitution ?? 0.5,
    friction: options.friction ?? 0.99,
    isStatic: isStatic,
  };
}

export function createLine(p1: Vec2, p2: Vec2, restitution: number = 0.8): Line {
  return {
    p1: { ...p1 },
    p2: { ...p2 },
    restitution,
  };
}

File: packages/server-rust/src/physics/body.rs
use super::Vec2;

#[derive(Clone)]
pub struct Circle {
    pub position: Vec2,
    pub velocity: Vec2,
    pub radius: f32,
    #[allow(dead_code)]
    pub mass: f32,
    pub inv_mass: f32,
    pub restitution: f32,
    pub friction: f32,
    pub is_static: bool,
}

impl Circle {
    pub fn new(
        position: Vec2,
        radius: f32,
        mass: f32,
        restitution: f32,
        friction: f32,
        is_static: bool,
    ) -> Self {
        let inv_mass = if is_static { 0.0 } else { 1.0 / mass };
        Self {
            position,
            velocity: Vec2::default(),
            radius,
            mass,
            inv_mass,
            restitution,
            friction,
            is_static,
        }
    }
}

#[derive(Clone)]
pub struct Line {
    pub p1: Vec2,
    pub p2: Vec2,
    pub restitution: f32,
}

impl Line {
    pub fn new(p1: Vec2, p2: Vec2, restitution: f32) -> Self {
        Self { p1, p2, restitution }
    }
}
